src/components/fields/file-upload/file-upload.tsx

```diff
@@ -0,0 +1,241 @@
+			maxFilesRuleRef.current = { max: lengthRule.length, errorMessage: lengthRule.errorMessage };
+		} else if (maxRule) {
+			maxFilesRuleRef.current = { max: maxRule.max, errorMessage: maxRule.errorMessage };
+		} else {
+			maxFilesRuleRef.current = {};
+		}
+
+		setFieldValidationConfig(
+			id,
+			Yup.array()
+				.of(
+					Yup.object().shape({
+						dataURL: Yup.string(),
+						fileId: Yup.string(),
+						fileName: Yup.string(),
+						fileUrl: Yup.string(),
+						uploadResponse: Yup.object(),
+					})
+				)
+				.test("is-empty-array", isRequiredRule?.errorMessage || ERROR_MESSAGES.UPLOAD().REQUIRED, (value) => {
+					if (!value || !isRequiredRule?.required) return true;
+					return value.length > 0;
+				})
+				.test(
+					"max-size-in-kb",
+					maxFileSizeRuleRef.current.errorMessage ||
+						ERROR_MESSAGES.UPLOAD().MAX_FILE_SIZE(maxFileSizeRuleRef.current.maxSizeInKb),
+					(value) => {
+						if (!value || !Array.isArray(value) || !maxFileSizeRuleRef.current.maxSizeInKb) return true;
+						return files
+							.filter((file) => file.status === EFileStatus.UPLOADED)
+							.every((file) => {
+								if (uploadOnAddingFile.type === "base64") {
+									return (
+										FileHelper.getFilesizeFromBase64(file.dataURL) <=
+										maxFileSizeRuleRef.current.maxSizeInKb * 1024
+									);
+								} else if (uploadOnAddingFile.type === "multipart") {
+									return file.rawFile.size <= maxFileSizeRuleRef.current.maxSizeInKb * 1024;
+								}
+							});
+					}
+				)
+				.test(
+					"max-files",
+					maxFilesRuleRef.current.errorMessage ||
+						ERROR_MESSAGES.UPLOAD().MAX_FILES(maxFilesRuleRef.current.max),
+					(value) => {
+						if (!value || !Array.isArray(value) || !maxFilesRuleRef.current.max) return true;
+						return value.length <= maxFilesRuleRef.current.max;
+					}
+				),
+			validation
+		);
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [validation, files.map((file) => file.status).join(",")]);
+
+	useEffect(() => {
+		// for defaultValue
+		if (!isDirty && Array.isArray(value)) {
+			const newFiles: IFile[] = [];
+			(value as IFileUploadValue[]).forEach(({ dataURL, fileId, fileName, fileUrl }, i) => {
+				newFiles.push({
+					addedFrom: "schema",
+					dataURL,
+					fileItem: {
+						id: fileId,
+					} as FileItemProps,
+					fileUrl,
+					rawFile: {
+						name: fileName,
+					} as File,
+					slot: i,
+					status: EFileStatus.INJECTED,
+				});
+			});
+			handleNewFiles(newFiles, []);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [isDirty]);
+
+	// =============================================================================
+	// EVENT HANDLERS
+	// =============================================================================
+	const handleChange = (rawFiles: File[]) => {
+		handleNewFiles(
+			rawFiles.map((rawFile) => ({
+				addedFrom: "input" as const,
+				rawFile,
+				slot: FileUploadHelper.findAvailableSlot(files),
```

Comment:
the same slot is assigned when multiple files are uploaded at the same time, as the files array is not the latest

anyway looking through the manager and component, I don't see the use of slot now as each file item now has an id?

Reply:
ok have fixed the slot assignment. 

the slots are from the `image-upload` component and it is used by the OS track, I am porting this feature over first